### ISSUE TITLE
3.0.110: segfaulting chunkserver

### ISSUE THREAD

**Issue by onlyjob** [moosefs/moosefs#342]
Apparently 3.0.110 introduced a bug that is causing chunkserver to segfault:

```
Feb 15 10:55:43  mfschunkserver[208865]: hdd_chunk_remove: CRC not flushed - writing now
Feb 15 10:55:43  kernel: [3120785.162901] mfschunkserver[101044]: segfault at 9 ip 000055e9b3adddec sp 00007fa1ebce9da0 error 6 in mfschunkserver[55e9b3acc000+53000]
```

Observed several times on one host, 65 times on another and about 300+ times on a third host since upgrade around 24 hours ago.

**Comment by onlyjob** [comment]
On both affected servers, segfaulting chunkservers are those where most new chunks are created as per storage classes definition of `-C`.

**Comment by pkonopelko** [comment]
Can you please send the output of:
```
addr2line -e /usr/lib/debug/usr/sbin/mfschunkserver 0x000055e9b3adddec
```
?

**Comment by onlyjob** [comment]
No such file. There are only numerical files under `/usr/lib/debug/.build-id`.

**Comment by pkonopelko** [comment]
Alright. I don't know what packaging you are using (whether it is your packaging from Salsa or maybe just built from sources by e.g. `linux_build.sh`), but the aim is to find a line of code where the segfault occurred. If it's your packaging from Salsa, `moosefs-chunkserver-dbgsym_*.deb` package is most probably needed to be installed first. If it's built from sources without packaging (and striping debug symbols), probably passing the path to the `mfschunkserver` binary as an argument to `-e` parameter should be fine.

**Comment by onlyjob** [comment]
Yes, I'm using build from Salsa for Debian and of course I have -dbgsym package installed yet I can't figure out the correct invocation of `addr2line` command.

**Comment by onlyjob** [comment]
I think the following commit is responsible: https://github.com/moosefs/moosefs/commit/de23226e7f10e2a4ea9790514d02c023e4c32329

**Comment by uu** [comment]
The same behaviour on gentoo. 

```
[8059057.152819] mfschunkserver[98770]: segfault at 9 ip 0000557f55f90122 sp 00007f08b8bfdda0 error 6 in mfschunkserver[557f55f7c000+52000]
[8059057.152838] Code: 48 8b 04 25 28 00 00 00 48 89 84 24 08 10 00 00 31 c0 e8 a1 c7 fe ff 85 c0 0f 85 69 01 00 00 48 8b 45 08 48 8d 3d fe 9a 05 00 <c6> 40 09 01 e8 25 c4 fe ff 41 89 c4 85 c0 0f 85 a5 00 00 00 0f b7
```

**Comment by acid-maker** [comment]
commit de23226e7f10e2a4ea9790514d02c023e4c32329 is ok. You shouldn't even see this `hdd_chunk_remove: CRC not flushed` message. It means that something is wrong - but definitely somewhere else.

Do you have in your log some other strange messages? Something like:

`serious error: crc changes lost` or something that look suspicious?

**Comment by onlyjob** [comment]
7313493 seems to help as there were no segfaults for about 12 hours. Thanks!
I'll test longer and report here.

There were nothing suspicious in chunkserver logs as far as I can tell...

**Comment by onlyjob** [comment]
So far definitely better but there was at least one new segfault in the chunkserver:

    segfault at 9 ip 00005585fc1d9dec sp 00007f1e11099da0 error 6 in mfschunkserver (deleted)[5585fc1c8000+53000]

**Comment by acid-maker** [comment]
Are you sure, that you run version with the patch? "mfschunkserver (deleted)" suggests that process was running from deleted (probably old) binary, not the current one. Maybe you only compiled new version and didn't restart.

**Comment by acid-maker** [comment]
I've added more protections just in case, but they are totally unnecessary in my opinion. Those protections are in code that worked fine for years.

BTW. Next time try to use addr2line - without it I can only guess what happened. More lines from syslog also helps.

**Comment by uu** [comment]
LGTM, bumped up on alpine and gentoo. Thanks! :)

**Comment by onlyjob** [comment]
Thanks. I tried my best to use `addr2line` but it never produce meaningful output... I think it is incompatible with -dbgsym symbols format.

**Comment by onlyjob** [comment]
Fixed in 3.0.111, thanks.
